helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java

```diff
@@ -425,25 +425,29 @@ private static boolean isJobComplete(JobContext ctx, Set<Integer> allPartitions,
         TaskPartitionState currState = TaskPartitionState.valueOf(instanceToCurrState.getValue());
         int pId = TaskUtil.getPartitionId(entry.getKey().getPartitionName());
 
-        // We must add all active task pIds back here because dropping transition could overwrite an
-        // active transition in paMap
-        // Add all task partitions in the following states:
-        // currState = INIT, requestedState = RUNNING (bootstrap)
-        // currState = RUNNING, requestedState = ANY (active)
-        // ** for tasks that are just in INIT state, we do not add them here because old
-        // Participants, upon connection reset, set tasks' currentStates to INIT. We cannot consider
-        // those tasks active **
-        if (result.containsKey(instance) && (currState == TaskPartitionState.INIT
-            && requestedState != null && requestedState.equals(TaskPartitionState.RUNNING.name())
-            || currState == TaskPartitionState.RUNNING)) {
-          // Check if this is a dropping transition
+        if (result.containsKey(instance)) {
+          // We must add all active task pIds back here because dropping transition could overwrite
+          // an active transition in paMap
+          // Add all task partitions in the following states:
+          // currState = INIT, requestedState = RUNNING (bootstrap)
+          // currState = RUNNING, requestedState = ANY (active)
+          // ** for tasks that are just in INIT state, we do not add them here because old
+          // Participants, upon connection reset, set tasks' currentStates to INIT. We cannot
+          // consider those tasks active **
+          if (currState == TaskPartitionState.INIT && requestedState != null
```

Comment:
Will this fix the COMPLETE with requested state is DROPPED?  I think COMPLETE/TASK_ERROR will not be added in map, right?

Reply:
See the next block where all tasks with their requested state is DROPPED get added to tasksToDrop.